### PR TITLE
Use access permissions instead of base_album table to determine access rights.

### PR DESCRIPTION
**Pull request by ildyria**
:warning: This change may induce slow-down.

- [x] : Migrate to access_permission table.
- [x] : Support sharing - visibility : behaviour is the same.
- [x] : Support visibility settings for Smart Albums => Most likely in settings to make it easier.
- [x] : Merge https://github.com/LycheeOrg/Lychee-front/pull/396

### CODE REVIEW

**Comment by ildyria** [comment]
Merging https://github.com/LycheeOrg/Lychee/pull/1795 will simplify this code change (especially the number of files).

**Comment by ildyria** [comment]
I still need to implement the front-end part, but at least server side is ready.

File: lang/de/lychee.php [hunk 1]
```diff
@@ -360,6 +360,10 @@
 	'LANG_TEXT' => 'Sprache für Lychee ändern:',
 	'LANG_TITLE' => 'Sprache festlegen',
 
+	'SETTING_RECENT_PUBLIC_TEXT' => 'Make "Recent" smart album accessible to anonymous users',
+	'SETTING_STARRED_PUBLIC_TEXT' => 'Make "Starred" smart album accessible to anonymous users',
+	'SETTING_ONTHISDAY_PUBLIC_TEXT' => 'Make "On This Day" smart album accessible to anonymous users',
```
Comment (qwerty287) [inline lang/de/lychee.php]:
```suggestion
	'SETTING_RECENT_PUBLIC_TEXT' => 'Intelligentes Album „Zuletzt benutzt“ für anonyme Nutzer aktivieren',
	'SETTING_STARRED_PUBLIC_TEXT' => 'Intelligentes Album „Favoriten“ für anonyme Nutzer aktivieren',
	'SETTING_ONTHISDAY_PUBLIC_TEXT' => 'Intelligentes Album „An diesem Tag“ für anonyme Nutzer aktivieren',
```

File: app/Actions/Album/Create.php [hunk 2]
```diff
@@ -29,16 +30,26 @@ public function create(string $title, ?Album $parentAlbum): Album
...
+		if ($defaultProtectionType === DefaultAlbumProtectionType::INHERIT && $parentAlbum !== null) {
+			$parentPermissions = $parentAlbum->access_permissions;
+			$copyPermissions = [];
+			foreach ($parentPermissions as $parentPermission) {
+				$copyPermissions[] = AccessPermission::ofAccessPermission($parentPermission);
+			}
+			// TODO: DOUBLE CHECK
+			// @phpstan-ignore-next-line
+			$album->access_permissions()->sync($copyPermissions);
+		}
```
Comment (qwerty287) [inline app/Actions/Album/Create.php]: Maybe we can put this into a helper in `AccessPermission`. Like `AccessPermission::ofAlbumDefault`.

File: app/Models/BaseAlbumImpl.php [hunk 3]
```diff
@@ -155,14 +153,8 @@ class BaseAlbumImpl extends Model implements HasRandomID
 		'owner_id' => 0,
 		'sorting_col' => null,
 		'sorting_order' => null,
-		// Security attributes
+		// "Security" attributes
```
Comment (qwerty287) [inline app/Models/BaseAlbumImpl.php]: Use a different name. Something like "Visibility"?

File: database/migrations/2023_04_19_052254_create_access_permissions.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	 */
+	public function down(): void
+	{
+		Schema::dropIfExists(self::TABLE_NAME);
+	}
+
+	/**
+	 * Generate the table for access_permissions.
+	 */
+	private function createAccesPersmissionTable(): void
```
Comment (qwerty287) [inline database/migrations/2023_04_19_052254_create_access_permissions.php]:
```suggestion
	private function createAccessPermissionTable(): void
```

Comment (ildyria) [reply database/migrations/2023_04_19_052254_create_access_permissions.php]: Fixed with [5180b35](https://github.com/LycheeOrg/Lychee/pull/1792/commits/5180b35d28ea3187693800406e7ef454bc753260).

File: app/Models/BaseAlbumImpl.php (hunk 3 above)
Comment (ildyria) [reply app/Models/BaseAlbumImpl.php]: Fixed with [5180b35](https://github.com/LycheeOrg/Lychee/pull/1792/commits/5180b35d28ea3187693800406e7ef454bc753260).

File: app/Actions/Album/Create.php (hunk 2 above)
Comment (ildyria) [reply app/Actions/Album/Create.php]: I had some thoughts and I don't think it is really useful.
However I extracted the code into an auxiliary function to clarify what is happening.
See [4963e0d](https://github.com/LycheeOrg/Lychee/pull/1792/commits/4963e0d42c56281f671f19557f2ecc9a983d9a41).

File: app/Http/Controllers/Administration/DiagnosticsController.php [hunk 5]
```diff
@@ -86,4 +92,44 @@ public function getSize(Space $space): array
 	{
 		return $this->isAuthorized() ? $space->get() : [self::ERROR_MSG];
 	}
+
+	public function getFullAccessPermissions(): View
```
Comment (qwerty287) [inline app/Http/Controllers/Administration/DiagnosticsController.php:101]: You could/should add a test to https://github.com/LycheeOrg/Lychee/blob/master/tests/Feature/DiagnosticsTest.php

File: app/Actions/Album/Create.php [hunk 6]
```diff
@@ -29,16 +30,21 @@ public function create(string $title, ?Album $parentAlbum): Album
...
+			// TODO: DOUBLE CHECK
+			// @phpstan-ignore-next-line
+			$album->access_permissions()->attach(AccessPermission::ofPublic());
+		}
+
+		if ($defaultProtectionType === DefaultAlbumProtectionType::INHERIT && $parentAlbum !== null) {
+			// TODO: DOUBLE CHECK
+			// @phpstan-ignore-next-line
+			$album->access_permissions()->sync($this->copyPermission($parentAlbum));
+		}
```
Comment (qwerty287) [inline app/Actions/Album/Create.php]: According to Codecov, these lines are untested. Maybe add some tests with changing config?

**Review by qwerty287 — APPROVED** [review]
With the code itself I'm fine, but I would add more tests. You can do this in a later PR too

File: app/Actions/Album/Create.php (hunk 6 above)
Comment (ildyria) [reply app/Actions/Album/Create.php]: Fixed in [d7dd4bf](https://github.com/LycheeOrg/Lychee/pull/1792/commits/d7dd4bfe39ca8255736fef56e5f300828c6e6afd).

**Pull request merged**